cpp/src/arrow/io/file.cc

```diff
@@ -118,33 +118,78 @@ namespace io {
 
 // ----------------------------------------------------------------------
 // Cross-platform file compatability layer
+
 #if defined(_MSC_VER)
+
 constexpr const char* kRangeExceptionError =
     "Range exception during wide-char string conversion";
+
+class PlatformFilename {
```

Comment:
No reason, that's a larger scope change than needed for this patch, though. Maybe @MaxRis can take a look. opened ARROW-1339